### DISCUSSION TITLE
Setting DataSource on RenderDocumentDirective

### DISCUSSION THREAD

**Discussion by abpatel** [stoiveyp/Alexa.NET.APL#92] [Q&A] [answered]
I am trying to figure out how to set the DataSource on the RenderDocumentDirective
`var launchDirective =
                        new RenderDocumentDirective(new
                        APLDocumentLink("doc://alexa/apl/documents/launchresponse")
                        );
 launchDirective.DataSources["key"]= ??? `
The value is a APLDataSource but I'm not sure how to set one. Is there an example that illustrates how to set DataSource?

**Comment by stoiveyp** [comment, marked as answer]
Hi @abpatel 

There are several different types of data sources, so it depends what your needs are. All the available data sources are in the namespace `Alexa.NET.APL.DataSources`

If you're looking to just set properties - similar to the example in:
https://developer.amazon.com/en-US/docs/alexa/alexa-presentation-language/apl-data-source.html#object-datasource

Then here the type is `object` so you need an ObjectDataSource, so your example would be something along the lines of....
```csharp
launchDirective.DataSources = new Dictionary<string, APLDataSource> {
    { "key", new ObjectDataSource{
          Properties = new Dictionary<string, object>{ { "key", "value" } }
    }
};
```

Hope this helps

**Reply by abpatel** [reply]
Awesome! I discovered that shortly before I saw your reply. I actually ended up using the KeyValueDataSource and it worked perfectly.
Are there any samples you might have that illustrate some of the typical usages of the apis in the library? 

Thank you for putting this library together.
Makes it so much more approachable for. Net devs to start building alexa skills.
